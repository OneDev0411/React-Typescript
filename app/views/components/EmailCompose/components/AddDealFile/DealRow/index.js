import React from 'react'

import { Avatar } from 'components/Avatar'
import { Container } from 'components/SelectContactModal/components/ContactItem'
import { getSide } from 'models/Deal/helpers/context'

import { Address, Details } from './styled'

export function DealRow({
  item,
  status,
  getAvatarProps,
  getPrice,
  onClickHandler,
  ...rest
}) {
  return (
    <Container {...rest} onClick={() => onClickHandler(item)}>
      <Avatar {...getAvatarProps(item)} />
      <div style={{ paddingLeft: '1em' }}>
        <Address>{item.title}</Address>
        <Details style={{ marginRight: '0.5em' }}>
          {status || 'Draft'} . {getSide(item)} . {item.property_type?.label}
        </Details>
      </div>
    </Container>
  )
}
