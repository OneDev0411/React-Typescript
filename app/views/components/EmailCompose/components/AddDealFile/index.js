import React, { Fragment } from 'react'

import AttachDealIcon from 'components/SvgIcons/AttachDeal/IconAttachDeal'
import SearchDealDrawer from 'components/SearchDealDrawer'
import SelectDealFileDrawer from 'components/SelectDealFileDrawer'

import { FooterAction } from '../FooterAction'
import { DealRow } from './DealRow'

export class AddDealFile extends React.Component {
  state = {
    isDealsListOpen: false,
    isDealFilesOpen: false,
    deal: this.props.deal
  }

  handleClick = () =>
    this.setState(state => ({
      isDealsListOpen: !state.deal,
      isDealFilesOpen: !!state.deal
    }))

  openDealDrawer = () =>
    this.setState({
      isDealsListOpen: true
    })

  closeDealDrawer = () =>
    this.setState({
      isDealsListOpen: false
    })

  closeDealFilesDrawer = () =>
    this.setState({
      deal: this.props.deal || null,
      isDealFilesOpen: false,
      isDealsListOpen: !this.props.deal
    })

  handleSelectDeal = deal => {
    this.setState({
      deal,
      isDealsListOpen: false,
      isDealFilesOpen: true
    })
  }

  handleChangeSelectedDealFile = files => {
    this.setState({
      deal: this.props.deal || null,
      isDealFilesOpen: false
    })

    const documents = files.map(file => ({
      ...file,
      attachmentType: 'deal-file'
    }))

    this.props.input.onChange(documents)
  }

  get InitialAttachments() {
    return this.props.initialAttachments.filter(
      item => item.attachmentType === 'deal-file'
    )
  }

  render() {
    return (
      <Fragment>
        <FooterAction tooltipCaption="Deal Files" onClick={this.handleClick}>
          <AttachDealIcon
            style={{
              width: '1.5rem',
              height: '1.5rem'
            }}
            fillColor="#000"
          />
        </FooterAction>

        {this.state.isDealsListOpen && (
          <SearchDealDrawer
            isOpen
            showBackdrop={false}
            title="Select a deal to view its files"
            itemRenderer={DealRow}
            onClose={this.closeDealDrawer}
            onSelect={this.handleSelectDeal}
          />
        )}

        {this.state.isDealFilesOpen && (
          <SelectDealFileDrawer
            isOpen
            title={`Select files from “${this.state.deal.title}”`}
            drawerOptions={{
              showBackdrop: false
            }}
            showStashFiles={false}
            initialAttachments={this.InitialAttachments}
            defaultSelectedItems={this.props.input.value}
            deal={this.state.deal}
            onChangeSelectedDocuments={this.handleChangeSelectedDealFile}
            onClose={this.closeDealFilesDrawer}
          />
        )}
      </Fragment>
    )
  }
}
