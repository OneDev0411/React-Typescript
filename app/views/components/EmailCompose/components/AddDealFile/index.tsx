import React, { MouseEventHandler, useState } from 'react'

import { ListItem } from '@material-ui/core'
import { mdiCurrencyUsdCircle } from '@mdi/js'
import { useSelector } from 'react-redux'

import SearchDealDrawer from 'components/SearchDealDrawer'
import SelectDealFileDrawer from 'components/SelectDealFileDrawer'
import { SvgIcon } from 'components/SvgIcons/SvgIcon'
import { useDeepMemo } from 'hooks/use-deep-memo'
import { IAppState } from 'reducers'
import { selectDealEnvelopes } from 'reducers/deals/envelopes'
import { selectDealTasks } from 'reducers/deals/tasks'
import { notUndefined } from 'utils/ts-utils'

import { getAllDealDocuments } from '../../../SelectDealFileDrawer/helpers/get-all-deal-documents'

import { DealRow } from './DealRow'

interface StateProps {
  checklists: Record<UUID, IDealChecklist>
  tasks: Record<UUID, IDealTask>
  envelopes: Record<UUID, IDealEnvelope>
}

interface Props {
  defaultSelectedDeal?: IDeal
  initialAttachments: IFile[]
  value: IFile[]
  onClick?: MouseEventHandler
  onChange: (files: IFile[]) => void
  onClickAddDealAttachments?: () => void
}

export function AddDealFile({
  initialAttachments,
  defaultSelectedDeal,
  onChange,
  onClick,
  onClickAddDealAttachments,
  value = []
}: Props) {
  const { checklists, tasks, envelopes }: StateProps = useSelector(
    ({ deals: { checklists, tasks, envelopes } }: IAppState) => ({
      checklists,
      tasks,
      envelopes
    })
  )
  const [isDealsListOpen, setDealsListOpen] = useState(false)
  const [isDealFilesOpen, setDealFilesOpen] = useState(false)
  const [deal, setDeal] = useState<IDeal | null>(defaultSelectedDeal || null)

  const handleClick = event => {
    if (defaultSelectedDeal) {
      onClick && onClick(event)
      onClickAddDealAttachments && onClickAddDealAttachments()

      return
    }

    setDealsListOpen(true)
    setDealFilesOpen(false)

    onClick && onClick(event)
  }

  const closeDealDrawer = () => setDealsListOpen(false)

  const closeDealFilesDrawer = () => {
    setDeal(deal || null)
    setDealsListOpen(true)
    setDealFilesOpen(false)
  }

  const handleSelectDeal = (deal: IDeal) => {
    setDeal(deal)
    setDealsListOpen(false)
    setDealFilesOpen(true)
  }

  const fileExists = (files: IFile[], file: IFile) =>
    files.some(aFile => aFile.id === file.id)

  const handleChangeSelectedDealFile = (files: IFile[]) => {
    setDeal(deal || null)
    setDealFilesOpen(false)

    if (deal) {
      const currentFiles: IFile[] = value || []

      // Previously selected files which are either non-deal files or
      // deal files that are still selected
      const preservedFiles = currentFiles.filter(
        file => !fileExists(allDealFiles, file) || fileExists(files, file)
      )

      // New deal files that are selected from the drawer, but were not
      // previously selected
      const newDealFiles = files.filter(file =>
        preservedFiles.every(aFile => aFile.id !== file.id)
      )

      onChange([...preservedFiles, ...newDealFiles])
    }
  }

  const allDealFiles: IDealFile[] =
    useDeepMemo(() => {
      return (
        (deal &&
          getAllDealDocuments(
            deal,
            selectDealEnvelopes(deal, envelopes),
            selectDealTasks(deal, checklists, tasks),
            true
          )) ||
        []
      )
    }, [initialAttachments, deal, envelopes, checklists, tasks]) || []

  const fileToDealFile = (file: IFile): IDealFile | undefined => {
    return allDealFiles.find(document => document.id === file.id)
  }

  const selectedDealFiles: IDealFile[] = (value || [])
    .map((file: IFile) => fileToDealFile(file))
    .filter(notUndefined)

  const initialDealFiles = (initialAttachments || [])
    .map(fileToDealFile)
    .filter(notUndefined)

  return (
    <>
      <ListItem button onClick={handleClick}>
        <SvgIcon path={mdiCurrencyUsdCircle} rightMargined />
        Attach from deals
      </ListItem>

      {isDealsListOpen && (
        <SearchDealDrawer
          isOpen
          title="Select a deal to view its files"
          itemRenderer={DealRow}
          onClose={closeDealDrawer}
          onSelect={handleSelectDeal}
        />
      )}

      {isDealFilesOpen && deal && (
        <SelectDealFileDrawer
          isOpen
          allowNoChecklist
          title={deal ? `Select files from "${deal.title}"` : 'Select files'}
          dealId={deal.id}
          initialAttachments={initialDealFiles}
          defaultSelectedItems={selectedDealFiles}
          drawerOptions={{
            hideBackdrop: true
          }}
          onChangeSelectedDocuments={handleChangeSelectedDealFile}
          onClose={closeDealFilesDrawer}
        />
      )}
    </>
  )
}

export default AddDealFile
