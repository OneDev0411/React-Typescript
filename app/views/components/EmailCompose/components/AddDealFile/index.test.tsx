import { ReactNode } from 'react'

import { fireEvent, render, RenderResult } from '@testing-library/react'
import { keyBy } from 'lodash'
import { FieldRenderProps } from 'react-final-form'
import { Provider } from 'react-redux'
import { createStore } from 'redux'

import deal1 from 'fixtures/deal/164ca424-6732-11e9-82bf-0a95998482ac.json'
import checklistsArray from 'fixtures/deal/checklists.json'
import dealWithDraft from 'fixtures/deal/deal-with-draft-files.json'
import deal2 from 'fixtures/deal/fa2126f8-6740-11e9-90e4-0a95998482ac.json'
import tasksArray from 'fixtures/deal/tasks.json'
import file1 from 'fixtures/files/file1.json'

import { TestBed } from '../../../../../../tests/unit/TestBed'

import AddDealFile, { AddDealFile as AddDealFileBase } from '.'

const checklists = keyBy(checklistsArray as IDealChecklist[], 'id')
const tasks = keyBy<IDealTask>(tasksArray as any, 'id')
const deals = [deal1, deal2, dealWithDraft] as any[]

const input: FieldRenderProps<any>['input'] = {
  name: 'test',
  onBlur: () => {},
  onFocus: () => {},
  onChange: () => {},
  value: [] as IFile[]
}

const AddDealFileTestBed = ({ children }: { children: ReactNode }) => {
  const dealsState = {
    list: keyBy(deals, 'id'),
    checklists,
    tasks,
    envelopes: {}
  }

  return (
    <TestBed
      reduxState={{
        deals: dealsState
      }}
    >
      {children}
    </TestBed>
  )
}

describe('AddDealFile component', () => {
  // https://gitlab.com/rechat/web/issues/2963
  test('renders', () => {
    const mockStore = createStore((state, action) => state, {
      deals: { checklists: {}, tasks: {}, envelopes: {} }
    })

    const container = render(
      <Provider store={mockStore}>
        <AddDealFileBase
          value={[]}
          onChange={input.onChange}
          initialAttachments={[]}
        />
      </Provider>
    )

    expect(container).toMatchSnapshot()
  })

  test('it selects deal files', () => {
    const onChange = jest.fn()
    const $ = render(
      <AddDealFileTestBed>
        <AddDealFile value={[]} onChange={onChange} initialAttachments={[]} />
      </AddDealFileTestBed>
    )

    toggleDealFiles($, '7 SW 7 Alley  Unit 7', ['Information About'])

    expect(onChange).toBeCalledWith([
      expect.objectContaining({
        id: '678d8af0-67fb-11e9-8a19-0a95998482ac'
      })
    ])
  })

  /**
   * https://gitlab.com/rechat/web/issues/3386
   */
  test('it preserves uploaded attachments when an attachment is added from deals', () => {
    const onChange = jest.fn()
    const $ = render(
      <AddDealFileTestBed>
        <AddDealFile
          value={[file1] as any}
          onChange={onChange}
          initialAttachments={[]}
        />
      </AddDealFileTestBed>
    )

    toggleDealFiles($, '7 SW 7 Alley  Unit 7', ['Information About'])

    expect(onChange).toBeCalledWith([
      file1,
      expect.objectContaining({
        id: '678d8af0-67fb-11e9-8a19-0a95998482ac'
      })
    ])
  })

  /**
   * https://gitlab.com/rechat/web/issues/3387
   */
  test('Unorganized (draft) files are suggested and selectable', () => {
    const onChange = jest.fn()
    const $ = render(
      <AddDealFileTestBed>
        <AddDealFile value={[]} onChange={onChange} initialAttachments={[]} />
      </AddDealFileTestBed>
    )

    toggleDealFiles($, '3508  Springbrook Street', ['Codal_p_588921_0.pdf'])

    expect(onChange).toBeCalledWith([
      expect.objectContaining({ id: dealWithDraft.files[0].id })
    ])
  })
})

function toggleDealFiles(
  $: RenderResult,
  dealName: string,
  dealFileNames: string[]
) {
  // Open deal selector drawer
  fireEvent.click($.getByText('Attach from deals'))

  // Select the deal. It will open deal file selector drawer
  fireEvent.click(
    $.getByText(dealName, {
      normalizer: text => text
    })
  )
  dealFileNames.forEach(dealFileName => {
    fireEvent.click($.getByText(dealFileName, { exact: false }))
  })
  fireEvent.click($.getByText('Next'))
}
