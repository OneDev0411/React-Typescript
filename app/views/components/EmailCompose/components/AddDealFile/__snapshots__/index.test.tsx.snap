// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddDealFile component renders 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        aria-disabled="false"
        class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
        role="button"
        tabindex="0"
      >
        <svg
          class="Icon-rightMargin-3"
          role="presentation"
          style="width: 1.5rem; height: 1.5rem;"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2A10 10 0 1 0 22 12A10 10 0 0 0 12 2M15 10H11V11H14A1 1 0 0 1 15 12V15A1 1 0 0 1 14 16H13V17H11V16H9V14H13V13H10A1 1 0 0 1 9 12V9A1 1 0 0 1 10 8H11V7H13V8H15Z"
            style="fill: currentColor;"
          />
        </svg>
        Attach from deals
        <span
          class="MuiTouchRipple-root"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      aria-disabled="false"
      class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
      role="button"
      tabindex="0"
    >
      <svg
        class="Icon-rightMargin-3"
        role="presentation"
        style="width: 1.5rem; height: 1.5rem;"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 2A10 10 0 1 0 22 12A10 10 0 0 0 12 2M15 10H11V11H14A1 1 0 0 1 15 12V15A1 1 0 0 1 14 16H13V17H11V16H9V14H13V13H10A1 1 0 0 1 9 12V9A1 1 0 0 1 10 8H11V7H13V8H15Z"
          style="fill: currentColor;"
        />
      </svg>
      Attach from deals
      <span
        class="MuiTouchRipple-root"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
